README.md

# kfa: K-Fold Cross-Validation For Factor Analysis

[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/kfa)](https://cran.r-project.org/package=kfa)
[![metacran
downloads](https://cranlogs.r-pkg.org/badges/kfa)](https://cran.r-project.org/package=kfa)
[![metacran
downloads](https://cranlogs.r-pkg.org/badges/grand-total/kfa)](https://cran.r-project.org/package=kfa)

**kfa** provides utilities for examining the dimensionality of a set of
variables to foster scale development. Harnessing a k-fold
cross-validation approach, **kfa** helps researchers compare possible
factor structures and identify which structures are plausible and
replicable across samples.

## Installation

``` r
# From CRAN
install.packages("kfa")

# Development version
install.packages("remotes")
remotes::install_github("knickodem/kfa")

library(kfa)
```

## Workflow

The two primary functions are `kfa()` and `kfa_report()`. When the set
of potential variables and (optionally) the maximum number of factors,
*m*, are supplied to `kfa()`, the function:

-   (if requested) conducts a power analysis to determine the number of
    folds, *k*, on which to split the data into training and testing
    samples
-   creates *k* folds (i.e. the training and testing samples).

Then for each fold:

-   calculates sample statistics (e.g., correlation matrix, thresholds
    \[if necessary\]) from training sample.
-   runs `2:m` factor exploratory factor analysis (EFA) models using the
    sample statistics, applies rotation (if specified), and extracts the
    factor structure for a confirmatory factor analysis (CFA). The
    structure for a 1-factor CFA is also defined.
-   runs the `1:m` factor CFA models on the testing sample.

The factor analyses are run using the `lavaan` package with many of the
`lavaan` estimation and missing data options available for use in
`kfa()`. `kfa()` returns a list of lists with *k* outer elements for
each fold and *m* inner elements for each replicable factor model, each
containing a `lavaan` object. To expedite running *k* x *m* x 2 (EFA and
CFA) models, the function utilizes the `parallel` and `foreach` packages
for parallel processing.

``` r
library(kfa)
# simulate data based on a 3-factor model with standardized loadings
sim.mod <- "f1 =~ .7*x1 + .8*x2 + .3*x3 + .7*x4 + .6*x5 + .8*x6 + .4*x7
                f2 =~ .8*x8 + .7*x9 + .6*x10 + .5*x11 + .5*x12 + .7*x13 + .6*x14
                f3 =~ .6*x15 + .5*x16 + .9*x17 + .4*x18 + .7*x19 + .5*x20
                f1 ~~ .2*f2
                f2 ~~ .2*f3
                f1 ~~ .2*f3
                x9 ~~ .2*x10"
set.seed(1161)
sim.data <- simstandard::sim_standardized(sim.mod,
                                          n = 900,
                                          latent = FALSE,
                                          errors = FALSE)[c(2:9,1,10:20)]

# include a custom 2-factor model
custom2f <- paste0("f1 =~ ", paste(colnames(sim.data)[1:10], collapse = " + "),
                   "\nf2 =~ ",paste(colnames(sim.data)[11:20], collapse = " + "))

mods <- kfa(data = sim.data,
            k = NULL,    # NULL prompts power analysis to determine number of folds
            custom.cfas = custom2f  # can be a single object or named list
            )
```

`kfa_report()` then aggregates the CFA model fit, parameter estimates,
and model-based reliability across folds for each factor structure
extracted in `kfa()`. The results are then organized and exported via
`rmarkdown`, such as the [example
report](https://htmlpreview.github.io/?https://github.com/knickodem/kfa/blob/main/README%20Example%20Reports/example_sim_kfa_report.html)
run below.

``` r
# Run report
kfa_report(models = mods,
           file.name = "example_sim_kfa_report",
           report.title = "K-fold Factor Analysis - Example Sim",
           report.format = "html_document")
```

## Under Development and Consideration

-   **Clustered Data** - The package does not currently account for
    clustered data. Future versions will utilize the cluster argument
    from `lavaan` to estimate cluster robust standard errors when
    calculating the correlation matrix for the factor analyses. We are
    also considering how to account for nesting structures in the
    creation of the folds, which are currently created assuming a simple
    random sample. If so, we will also incorporate cluster adjustments
    for the power analysis determining the value of *k*.


NEWS.md
# kfa 0.2.2

*New features and adjustments:
  * loading flag based on absolute value of loading
  * The degrees of freedom calculation in `find_k` is now based on a CFA model rather than EFA
  * Can toggle plots off when generating report - temporarily changed default to off to limit strong dependency with archived package

# kfa 0.2.1

*New features and adjustments:
  * `data` argument added and `variable` argument re-defined as specifying column names in `data`
  * The `sampling.weights` argument is passed appropriately to `lavaan` and incorporated into the analysis
  * Changed name of `threshold` argument to the more specific name `min.loading`
  * Adjusted order of `custom.cfa` models in flag table to match the order in the model fit table of the report

# kfa 0.2.0

* New features and major adjustments:
  * Users can keep cross-loading variables by setting `simple = FALSE` and specifying a `threshold`
  * Default test statistic and fit indices switched from naive to scaled versions (mean and variance adjusted)
  * In `agg_model_fit`, changed default index to `"all"`
  * `model_structure`, `get_std_loadings`, and `index_available` are exported for use
  * Tables sorted based on order of variables and models provided by user rather than alphabetical
* Bugs and background
  * Deleted Rd files for non-exported functions - silly me :)
  * Fixed rotation error bug in `run_efa`
  * path disassociated with report.title
  * Plot edges are set to black

# kfa 0.1.0

* Initial CRAN submission


R/kfa_report.R
#' Creates summary report from a k-fold factor analysis
#'
#' Generates a report summarizing the factor analytic results over k-folds.
#'
#' @param models an object returned from \code{\link[kfa]{kfa}}
#' @param file.name character; file name to create on disk.
#' @param report.title character; title of the report
#' @param path character; path of the directory where summary report will be saved. Default is working directory. \code{path} and \code{file.name} are combined to create final file path
#' @param report.format character; file format of the report. Default is HTML ("html_document"). See \code{\link[rmarkdown]{render}} for other options.
#' @param word.template character; file path to word document to use as a formatting template when \code{report.format = "word_document"}.
#' @param index character; one or more fit indices to summarize in the report. Use \code{\link[kfa]{index_available}} to see choices.
#' Chi-square value and degrees of freedom are always reported. Default is CFI and RMSEA (naive, scaled, or robust version depends on estimator used in \code{models}).
#' @param plots logical; should plots of the factor models be included in the report?
#' @param load.flag numeric; factor loadings of variables below this value will be flagged. Default is .30
#' @param cor.flag numeric; factor correlations above this value will be flagged. Default is .90
#' @param rel.flag numeric; factor (scale) reliabilities below this value will be flagged. Default is .60.
#' @param digits integer; number of decimal places to display in the report.
#'
#' @return A summary report of factor structures and model fit within and between folds.
#'
#' @examples
#'
#' # simulate data based on a 3-factor model with standardized loadings
#' sim.mod <- "f1 =~ .7*x1 + .8*x2 + .3*x3 + .7*x4 + .6*x5 + .8*x6 + .4*x7
#'                 f2 =~ .8*x8 + .7*x9 + .6*x10 + .5*x11 + .5*x12 + .7*x13 + .6*x14
#'                 f3 =~ .6*x15 + .5*x16 + .9*x17 + .4*x18 + .7*x19 + .5*x20
#'                 f1 ~~ .2*f2
#'                 f2 ~~ .2*f3
#'                 f1 ~~ .2*f3
#'                 x9 ~~ .2*x10"
#' set.seed(1161)
#' sim.data <- simstandard::sim_standardized(sim.mod, n = 900,
#'                                           latent = FALSE,
#'                                           errors = FALSE)[c(2:9,1,10:20)]
#'
#' # include a custom 2-factor model
#' custom2f <- paste0("f1 =~ ", paste(colnames(sim.data)[1:10], collapse = " + "),
#'                    "\nf2 =~ ",paste(colnames(sim.data)[11:20], collapse = " + "))
#'
#' \donttest{
#' mods <- kfa(data = sim.data,
#'             k = NULL, # prompts power analysis to determine number of folds
#'             cores = 2,
#'             custom.cfas = custom2f)
#'             }
#'
#' \dontrun{
#' kfa_report(mods, file.name = "example_sim_kfa_report",
#'            report.format = "html_document",
#'            report.title = "K-fold Factor Analysis - Example Sim")
#'            }
#'
#' @import lavaan
#' @import rmarkdown
#' @import flextable
#' @importFrom knitr opts_chunk
#' @importFrom knitr knit
#' @importFrom knitr knit_print
#' @importFrom officer fp_border
#' @importFrom simstandard sim_standardized
#'
#' @export

kfa_report <- function(models,
                       file.name, report.title = file.name,
                       path = NULL,
                       report.format = "html_document",
                       word.template = NULL,
                       index = "default",
                       plots = FALSE,
                       load.flag = .30, cor.flag = .90, rel.flag = .60,
                       digits = 2){

  if(inherits(models, "kfa")){
    cfas <- models$cfas
  } else {
    stop("models must be of class 'kfa'.")
  }

  ## analysis summary info
  k <- length(cfas) # number of folds
  m <- max(unlist(lapply(cfas, length))) # number of models per fold (includes both efa AND custom structures); m == length(mnames)
  mnames <- models$model.names # model names
  fac.allow <- length(models$efa.structures)
  fac.max <- max(as.numeric(substring(mnames[grepl("-factor", mnames)], 1, 1)))  # kfa naming convention "#-factor"; custom functions are assumed to have different convention
  vnames <- dimnames(lavaan::lavInspect(cfas[[1]][[1]], "sampstat")$cov)[[1]] # variable names
  nvars <- length(vnames)
  nobs <- sum(unlist(lapply(cfas, function(x) lavaan::lavInspect(x[[1]], "nobs"))))
  opts <- lavaan::lavInspect(cfas[[1]][[1]], "options") # estimation options; assumed to be the same for all models

  #### Model Fit ####
  ## summarizing fit statistics by fold
  kfits <- k_model_fit(models, index = index, by.fold = TRUE) # dataframe for each fold
  fit.table <- agg_model_fit(kfits, index = "all", digits = 2)
  # adjust model order to match model.names and other output
  fit.table <- fit.table[order(factor(fit.table$model, levels = mnames)),]

  ## best model in each fold
  # best.model <- best_model(kfits, index = index)

  ## creating appendix -  folds x model table of fit statistics
  mfits <- k_model_fit(models, index = index, by.fold = FALSE)
  appendix <- get_appendix(mfits, index = "all")

  #### Parameters ####
  ## model structures
  kstructures <- model_structure(models)

  ## loadings
  klambdas <- agg_loadings(models, flag = load.flag, digits = digits)

  ## factor correlations
  kcorrs <- agg_cors(models, flag = cor.flag)

  ## score reliabilities
  krels <- agg_rels(models, flag = rel.flag, digits = digits)

  ## flagged problems
  flagged <- model_flags(models, kstructures, klambdas, kcorrs, krels)



  ## running report
  if(report.format == "word_document"){
    width <- 6.5
  } else {
    width <- NULL
  }

  if(is.null(word.template)){
    word.template <- system.file("rmd", "kfa_word_template.docx", package = "kfa")
  }

  if(is.null(path)){
    path <- getwd()
  }
  report.title <- report.title # Unless already specified by user, sets report.title = file.name b/f path gets added
  file.name <- file.path(path, file.name)
  template <- system.file("rmd", "kfa-report.Rmd", package = "kfa")
  rmarkdown::render(input = template,
                    output_format = report.format,
                    output_file = file.name,
                    output_options = list(toc = TRUE, toc_depth = 2,
                                          always_allow_html = TRUE,
                                          reference_docx = word.template))
}


#' Default flextable format
#'
#' Internal function for formatting flextables
#'
#' @param df a \code{data.frame}
#' @param bold.type character indicating table with a pre-specified bolding pattern. Not currently implemented.
#' @param width numeric; maximum width of table in inches.
#' @param digits integer; number of decimal places to display in the report.
#'
#' @return a \code{flextable} object
#'
#' @import flextable
#'
#' @noRd

flextab_format <- function(df, bold.type = "none", width = NULL, digits = 2){

  numericcols <- which(unlist(lapply(df, is.numeric)))

  flex <- flextable::flextable(df)
  flex <- flextable::colformat_double(flex, j = numericcols, digits = digits)
  flex <- flextable::align(flex, i = NULL, j = NULL, align = "center", part = "all")
  flex <- flextable::font(flex, fontname = "Times New Roman", part = "all")
  flex <- flextable::padding(flex, padding = 3, part = "all")

  if(bold.type == "fit"){
    flex <- bold(flex, i = ~rmsea == min(rmsea), part =  "body")
  } else if(bold.type == "lambda"){
    flex <- bold(flex, i = ~mean < .3, part = "body")
    # error occurs when .3 is replaced by cut; not sure why
  }

  if(!is.null(width)){
    flex <- flextable::fit_to_width(flex, max_width = width)
  }
  flex <- flextable::autofit(flex)

  return(flex)
}

#' Default flextable format for header with two levels
#'
#' Internal function for formatting flextables
#'
#' @param flex a \code{flextable} object
#' @param mapping a \code{data.frame} specifying header columns. See \code{\link[flextable]{set_header_footer_df}} for details.
#' @param vert.cols columns where level 1 and level 2 cells will be vertically merged. See \code{\link[flextable]{merge_v}} for details.
#' @param border format of of horizontal borders. See \code{\link[flextable]{border_inner_h}} for details.
#'
#' @return a \code{flextable} object
#'
#' @import flextable
#'
#' @noRd

two_level_flex <- function(flex, mapping, vert.cols, border){

  flex <- flextable::set_header_df(flex, mapping = mapping)
  flex <- flextable::merge_h(flex, part = "header")
  flex <- flextable::merge_v(flex, j = vert.cols, part = "header")
  flex <- flextable::fix_border_issues(flex)
  flex <- flextable::border_inner_h(flex, border = border, part = "header")
  flex <- flextable::hline_top(flex, border = border, part = "all")
  # flex <- flextable::theme_vanilla(flex)
  flex <- flextable::align(flex, align = "center", part = "all")
  flex <- flextable::font(flex, fontname = "Times New Roman", part = "all")
  flex <- flextable::padding(flex, padding = 3, part = "all")
  flex <- flextable::autofit(flex)
  return(flex)
}


#' Create appendix table
#'
#' Internal function for converting appendix from \code{data.frame} to \code{flextable}
#'
#' @param appendix object returned from \code{\link[kfa]{get_appendix}}
#' @param mapping a \code{data.frame} specifying header columns. See \code{\link[flextable]{set_header_footer_df}} for details.
#' @param border format of of horizontal borders. See \code{\link[flextable]{border_inner_h}} for details.
#' @param digits integer; number of decimal places to display in the report.
#'
#' @return a \code{flextable} object
#'
#' @import flextable
#'
#' @noRd

appendix_wrapper <- function(appendix, mapping, border, digits){

  appendix.flex <- flextable::flextable(appendix)
  appendix.flex <- flextable::colformat_double(appendix.flex, j = -c(1), digits = digits)
  appendix.flex <- two_level_flex(appendix.flex, mapping = mapping, vert.cols = c("fold"), border = border)
  return(appendix.flex)

}
